File: python/vrml/isa.py
"""Instruction set for the Vector-Rotational Machine Language (VRML).

The ISA is deliberately tiny: every opcode is a geometric verb
(translation, rotation, projection, intersection, composition, comparison).
Higher-level constructs (arithmetic, branching, memory, subroutines) are
derived from these primitives.

Operand conventions:
    A, B          : vector registers
    S             : scalar register (float; integer values are encoded as
                   rounded floats -- the VM has no integer type, by design,
                   because integers are a special case of scalars under the
                   same geometric interpretation.)
    R0..R7        : rotor (transformation) registers
    P             : current program rotor (read-only for the program)
    C             : state / cursor vector
    F             : flag vector (gt, eq, lt) from the last CMP
    addr          : memory address (non-negative integer)
    k             : axis label (string: "X", "Y", "Z", "XI", "YI", "ZI", etc.)
    target        : label (string)
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Tuple, Union, Optional

# operand types
Operand = Union[str, int, float, Tuple[float, float, float]]


@dataclass(frozen=True)
class Instr:
    opcode: str
    args: Tuple[Operand, ...] = ()

    def __repr__(self) -> str:
        if not self.args:
            return f"{self.opcode}"
        return f"{self.opcode} " + ", ".join(_fmt(a) for a in self.args)


def _fmt(a):
    if isinstance(a, str):
        return a
    if isinstance(a, tuple):
        return "(" + ", ".join(repr(x) for x in a) + ")"
    return repr(a)


# The full opcode table.  Keep these stable; the assembler keys on them.
OPCODES = {
    # vector / scalar data movement
    "LOAD_AXIS",     # LOAD_AXIS k            -> A = unit(k)
    "LOAD_VEC",      # LOAD_VEC x,y,z         -> A = Vec(x,y,z)
    "LOADI",         # LOADI s                -> S = float(s)
    "MOV",           # MOV dst, src           -> dst = src
    "OUTI",          # OUTI S                 -> emit scalar S
    "OUTV",          # OUTV A|C               -> emit vector

    # geometric primitives -- the heart of the machine
    "TRANSLATE",     # TRANSLATE C|A, v       -> dst += v       (translation)
    "ROTATE",        # ROTATE C|A, Rk         -> dst = Rk.apply(dst)   (rotation)
    "ROTATE_AXIS",   # ROTATE_AXIS dst, k, theta  -> dst = R(k, theta).apply(dst)
    "COMPOSE",       # COMPOSE Rd, Rs         -> Rd = Rd * Rs   (compose rotors)
    "INVERSE",       # INVERSE Rk             -> Rk = Rk.inverse()
    "PROJ",          # PROJ A, B, t           -> A = lerp(A, B, t) (project)
    "INTERSECT",     # INTERSECT A, B         -> A = A ^ B (bivector)
    "SCALE",         # SCALE A, s             -> A = s * A
    "NORM",          # NORM A                 -> S = A.norm()
    "INNER",         # INNER A, B             -> S = A * B (dot)
    "CROSS",         # CROSS A, B             -> A = A.cross(B)
    "DUAL",          # DUAL A                 -> A = I * A  (vec <-> bivec)

    # comparison / branching (control flow as geometry)
    "CMP",           # CMP A, B               -> F = sign(A-B)
    "BRANCH_GT",     # BRANCH_GT target       -> if F.x>0: jump
    "BRANCH_EQ",     # BRANCH_EQ target       -> if F.y>0: jump
    "BRANCH_LT",     # BRANCH_LT target       -> if F.z>0: jump
    "BRANCH_AXIS",   # BRANCH_AXIS k, target  -> if C projected on k > 0: jump
    "JMP",           # JMP target             -> unconditional jump
    "CALL",          # CALL target            -> push frame, jump
    "RET",           # RET                    -> pop frame
    "HALT",          # HALT                   -> stop

    # memory (addresses are coordinates in a 1D lattice)
    "STORE",         # STORE A|S, addr        -> mem[addr] = src
    "LOAD",          # LOAD A, addr           -> A = mem[addr]
}


# fixed axis table for symbolic LOAD_AXIS / ROTATE_AXIS
AXES = {
    "X":  (1.0, 0.0, 0.0),
    "Y":  (0.0, 1.0, 0.0),
    "Z":  (0.0, 0.0, 1.0),
    "XI": (-1.0, 0.0, 0.0),
    "YI": (0.0, -1.0, 0.0),
    "ZI": (0.0, 0.0, -1.0),
}


def axis_vec(name: str):
    from .algebra import Vec
    return Vec(*AXES[name])

File: src/isa.rs
//! BT-IS instruction set.
//!
//! Every opcode is a geometric verb acting on cube state. Opcodes are
//! encoded as small integers (`u8`) and dispatched in O(1).
//!
//! ## Opcode map
//!
//! ```text
//!   0..32   : data movement, I/O, register ops
//!  32..56   : rotations applied to C
//!  56..64   : reflections / inversion / rotor composition
//!  64..96   : arithmetic on the ternary scalar encoded in C.x
//!  96..128  : comparison, three-way branching
//! 128..160  : memory (cube-addressed)
//! 192..     : meta (HALT = 192)
//! ```
//!
//! ## Register file
//!
//! - `C`       : cube (the program's cursor / data)
//! - `F`       : cube (last comparison sign, as `(gt, eq, lt)`)
//! - `R0..R7`  : rotor registers (composable permutations)
//! - `IP`      : instruction pointer
//! - `stack`   : return IPs for CALL
//! - `mem`     : `HashMap<Cube, Cube>` keyed by full cube coordinate

pub mod opcodes {
    // 0..32 : data / I/O / register ops
    pub const NOP: u8 = 0;
    pub const LOADC: u8 = 1;        // C := Cube(arg, arg, arg)
    pub const LOAD_AXIS: u8 = 2;    // C := unit_axis(arg)
    pub const OUTI: u8 = 3;         // emit C.x
    pub const OUTV: u8 = 4;         // emit C
    pub const MOV_R: u8 = 5;        // R[dst] := R[src]  (arg = dst, target = src)
    pub const LOAD_R: u8 = 6;       // R[arg] := Perm::identity()

    // 32..56 : rotations applied to C
    pub const ROT_Z_90:  u8 = 32;
    pub const ROT_Z_180: u8 = 33;
    pub const ROT_Z_270: u8 = 34;
    pub const ROT_X_90:  u8 = 35;
    pub const ROT_X_180: u8 = 36;
    pub const ROT_X_270: u8 = 37;
    pub const ROT_Y_90:  u8 = 38;
    pub const ROT_Y_180: u8 = 39;
    pub const ROT_Y_270: u8 = 40;
    pub const APPLY_R:   u8 = 41;   // C := R[arg].apply(C)

    // 56..64 : reflections / inversion / rotor composition
    pub const REFLECT_X:  u8 = 56;
    pub const REFLECT_Y:  u8 = 57;
    pub const REFLECT_Z:  u8 = 58;
    pub const NEG:        u8 = 59;
    pub const COMPOSE_R:  u8 = 60;  // R[arg] := R[arg] * R[target]
    pub const INVERSE_R:  u8 = 61;  // R[arg] := R[arg].inverse()

    // 64..96 : arithmetic (saturating balanced-ternary on C.x)
    pub const IADD: u8 = 64;
    pub const ISUB: u8 = 65;
    pub const IMUL: u8 = 66;

    // 96..128 : comparison, three-way branching
    pub const CMP:     u8 = 96;
    pub const BR_NEG:  u8 = 97;
    pub const BR_ZERO: u8 = 98;
    pub const BR_POS:  u8 = 99;
    pub const JMP:     u8 = 100;
    pub const CALL:    u8 = 101;
    pub const RET:     u8 = 102;
    pub const BR_AXIS: u8 = 103;

    // 128..160 : memory (cube-addressed)
    pub const STORE:    u8 = 128;   // mem[Cube(arg,arg,arg)] := C    (axial shortcut)
    pub const LOAD:     u8 = 129;   // C := mem[Cube(arg,arg,arg)]
    pub const STORE_C:  u8 = 130;   // mem[C] := C                    (full-cube address)
    pub const LOAD_C:   u8 = 131;   // C := mem[C]

    // 192.. : meta
    pub const HALT: u8 = 192;
}

/// Number of rotor registers available to programs.
pub const ROTOR_COUNT: usize = 8;
